src/main/java/net/minecraftforge/common/lighting/LightingEngine.java

```diff
@@ -0,0 +1,630 @@
+
+            if (newLight > this.neighborsLight[index])
+            {
+                this.enqueueBrightening(nPos, posToLong(nPos), newLight, nChunk);
+            }
+        }
+    }
+
+    private void enqueueBrighteningFromCur(final int newLight)
+    {
+        this.enqueueBrightening(this.curPos, this.curData, newLight, this.curChunk);
+    }
+
+    /**
+     * Enqueues the pos for brightening and sets its light value to <code>newLight</code>
+     */
+    private void enqueueBrightening(final BlockPos pos, final long longPos, final int newLight, final Chunk chunk)
+    {
+        this.queuedBrightenings[newLight].add(longPos);
+        chunk.setLightFor(this.lightType, pos, newLight);
+    }
+
+    /**
+     * Enqueues the pos for darkening and sets its light value to 0
+     */
+    private void enqueueDarkening(final BlockPos pos, final long longPos, final int oldLight, final Chunk chunk)
+    {
+        this.queuedDarkenings[oldLight].add(longPos);
+        chunk.setLightFor(this.lightType, pos, 0);
+    }
+
+    private static BlockPos longToPos(final MutableBlockPos pos, final long longPos)
+    {
+        final int posX = (int) (longPos >> sX & mX) - (1 << lX - 1);
+        final int posY = (int) (longPos >> sY & mY);
+        final int posZ = (int) (longPos >> sZ & mZ) - (1 << lZ - 1);
+        return pos.setPos(posX, posY, posZ);
+    }
+
+    private static long posToLong(final BlockPos pos)
+    {
+        return posToLong(pos.getX(), pos.getY(), pos.getZ());
+    }
+
+    private static long posToLong(final long x, final long y, final long z)
+    {
+        return (z + (1 << lZ - 1) << sZ) | (y << sY) | (x + (1 << lX - 1) << sX);
+    }
+
+    /**
+     * Polls a new item from <code>curQueue</code> and fills in state data members
+     *
+     * @return If there was an item to poll
+     */
+    private boolean nextItem()
+    {
+        if (this.curQueue.isEmpty())
+        {
+            return false;
+        }
+
+        this.curData = this.curQueue.poll();
+        this.isNeighborDataValid = false;
+        longToPos(this.curPos, this.curData);
+        this.curChunk = this.curToChunk();
+
+        return true;
+    }
+
+    private int posToCachedLight(final MutableBlockPos pos, final Chunk chunk)
+    {
+        return chunk.getCachedLightFor(this.lightType, pos);
+    }
+
+    private int curToCachedLight()
+    {
+        return this.posToCachedLight(this.curPos, this.curChunk);
+    }
+
+    /**
+     * Calculates the luminosity for <code>curPos</code>, taking into account <code>lightType</code>
+     */
+    private int curToLuminosity(final IBlockState state)
+    {
+        return this.lightType == EnumSkyBlock.SKY
+            ? this.curChunk.canSeeSky(this.curPos)
+            ? EnumSkyBlock.SKY.defaultLightValue
+            : 0
+            : MathHelper.clamp(state.getLightValue(this.world, this.curPos), 0, MAX_LIGHT);
```

Comment:
nested ternaries are harder to mentally parse than if statements and they take about the same lines :P